src/mame/drivers/myb3k.cpp

```diff
@@ -77,173 +91,426 @@ class myb3k_state : public driver_device
+	}
+}
 
-	for(y = 0; y < mc6845_v_display; y++)
+MC6845_UPDATE_ROW( myb3k_state::crtc_update_row )
+{
+	/* The 6845 is not programmed to get 80 character modes or 400 pixels width but this is managed by external circuitry that selects
+	   the apropriate pixelclock based on the video mode latch at i/o address 0x04. This callback always get x_count set to 40  */
+	uint8_t row_length = (m_vmode & 0x02) || (m_vmode == 0 && (m_pal == &m_mpal)) ? x_count * 2 : x_count;
+	for (int x_pos = 0; x_pos < row_length; x_pos++)
 	{
-		for(x = 0; x < mc6845_h_display; x++)
+		uint16_t page = (m_portc & PC1_SETPAGE) ? 0x8000 : 0;
+
+		if ((m_portc & PC2_DISPST) == 0)
+		{
+			for (int pxl = 0; pxl < 8; pxl++)
+			{
+				bitmap.pix32(y, ( x_pos * 8) + pxl) = rgb_t::black();
+			}
+		}
+		else
 		{
-			/* 8x8 grid gfxs, weird format too ... */
-			for(yi=0;yi<mc6845_tile_height;yi++)
+		  //uint32_t pdat;
+			switch (m_vmode)
 			{
-				for(xi=0;xi<8;xi++)
+			case 0:
+				// Color connector doing 320x200, 40x25 characters, 8 grey tones
+				if ((ioport("MONITOR")->read() & 1) == 0)
 				{
-					dot = (m_p_vram[(x+y*mc6845_h_display)*h_step+yi+0x8000] >> (7-xi)) & 1;
-					if((yi & ~7 && (!(m_vmode & 4))) || (yi & ~0xf && (m_vmode & 4)))
-						dot = 0;
+					uint32_t rowstart = (((x_pos + ma) * 32 + ra) & 0x7fff) + page;
+					uint32_t pdat  = ((m_vram[rowstart +  0]  & 0xff) << 16); // Green 8 bits
+					pdat |= ((m_vram[rowstart +  8]  & 0xf0) << 8);  // Red upper 4 bits
+					pdat |= ((m_vram[rowstart +  8]  & 0x0f) <<  4); // Blue upper 4 bits
+					pdat |= ((m_vram[rowstart +  24] & 0xf0) <<  4); // Red lower 4 bits
+					pdat |= ((m_vram[rowstart +  24] & 0x0f) <<  0); // Blue lower 4 bits
+					for (int pxl = 0; pxl < 8; pxl++)
+					{
+						uint16_t pind = 0;
+						/* Pixeldata for 8 pixels are stored in pdat as GGRRBB  */
+						pind = (((pdat & (0x800000 >> pxl)) ? 0x04 : 0x00) |
+							((pdat & (0x008000 >> pxl)) ? 0x02 : 0x00) |
+							((pdat & (0x000080 >> pxl)) ? 0x01 : 0x00) );
+
+						/* Check if we are in a cursor and create cursor if so */
+						pind ^= ((cursor_x != -1 && x_pos == cursor_x && ra == 7) ? 7 : 0);
+
+						/* Create the grey scale */
+						bitmap.pix32(y, ( x_pos * 8) + pxl) = rgb_t(m_mpal[pind & 0x07].g(),m_mpal[pind & 0x07].g(),m_mpal[pind & 0x07].g());
+					}
+				}
+				else // Monochrome connector doing 640x200, 80x25, b&w
+				{
+					uint32_t rowstart = (((x_pos + ma * 2) * 16 + ra) & 0x7fff) + page;
+					uint16_t pdat = m_vram[rowstart];
+					for (int pxl = 0; pxl < 8; pxl++)
+					{
+						if ((pdat & (0x80 >> pxl)) != 0)
+						{
+							//bitmap.pix32(y, ( x_pos * 8) + pxl) = (*m_pal)[0x07];
+							bitmap.pix32(y, ( x_pos * 8) + pxl) = m_mpal[0x07];
+						}
+						else
+						{
+							bitmap.pix32(y, ( x_pos * 8) + pxl) = rgb_t::black();
+						}
+					}
+				}
+				break;
+			case 1:
+				{
+					uint32_t pdat;
+					// Crude fix of 36CH mode
+					if ((ioport("DSW1")->read() & 0x0c) == 0 && ra > 7)
```

Comment:
This is a runtime tagmap lookup, which causes poor performance and makes refactoring harder.  Use `required_ioport` instead.